keysign/bluetoothoffer.py

```diff
@@ -0,0 +1,143 @@
+import logging
+from bluetooth import BluetoothSocket, RFCOMM
+import dbus
+import select
+import socket
+if __name__ == "__main__":
+    import gi
+    gi.require_version('Gtk', '3.0')
+    from twisted.internet import gtk3reactor
+    gtk3reactor.install()
+    from twisted.internet import reactor
+from twisted.internet import threads
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+if __name__ == "__main__" and __package__ is None:
+    logging.getLogger().error("You seem to be trying to execute " +
+                              "this script directly which is discouraged. " +
+                              "Try python -m instead.")
+    import os
+    parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    os.sys.path.insert(0, parent_dir)
+    os.sys.path.insert(0, os.path.join(parent_dir, 'monkeysign'))
+    import keysign
+    __package__ = str('keysign')
+
+from .gpgmh import get_public_key_data, get_usable_keys
+from .util import get_local_bt_address, mac_generate
+
+log = logging.getLogger(__name__)
+
+
+class BluetoothOffer:
+    def __init__(self, key, port=3, size=1024):
+        self.key = key
+        self.port = port
+        self.size = size
+        self.server_socket = None
+        self.message_def = None
+        self.stopped = False
+        self.code = None
+
+    @inlineCallbacks
+    def start(self):
+        self.stopped = False
+        message = "Back"
+        success = False
+        try:
+            while not self.stopped and not success:
+                # server_socket.accept() is not stoppable. So with select we can call accept()
+                # only when we are sure that there is already a waiting connection
+                ready_to_read, ready_to_write, in_error = yield threads.deferToThread(
+                    select.select, [self.server_socket], [], [], 0.5)
+                if ready_to_read:
+                    # We are sure that a connection is available, so we can call
+                    # accept() without deferring it to a thread
+                    client_socket, address = self.server_socket.accept()
+                    key_data = get_public_key_data(self.key.fingerprint)
+                    kd_decoded = key_data.decode('utf-8')
+                    yield threads.deferToThread(client_socket.sendall, kd_decoded)
```

Comment:
We could probably get away with setting the socket into nonblocking mode and then sendall. Then we'd save the overhead of a thread and we'd probably eliminate twisted here altogether.
Can you experiment around whether we can shave off this twisted thing here?
But what I wanted to get at was that I have the feeling that it's slightly cleaner to close() the client socket. That's currently not done, right? On purpose? Otherwise I suggest to call close().

Reply:
>     I'm talking about the client_socket
> 
> Ops, I was looking at server_socket :smile:

Heh. will you do a `client_socket,shutdown()` then?

Reply:
Added with 540ec80